File: 10-cons/cons.py
#Problem 10: Consensus and Profile
#Given: A collection of at most 10 DNA strings of equal length (at most 1 kbp) in FASTA format.
#Return: A consensus string and profile matrix for the collection. (If several possible consensus strings exist, then you may return any one of them.)

'''
A matrix is a rectangular table of values divided into rows and columns. An m×n matrix has m rows and
n columns. Given a matrix A, we write Ai,j to indicate the value found at the
intersection of row i and column j.

Say that we have a collection of DNA strings, all having the same length n.
Their profile matrix is a 4×n matrix P in which P1,j represents the number of times
that 'A' occurs in the jth position of one of the strings, P2,j represents the number of times
that C occurs in the jth position, and so on (see below).

A consensus string c is a string of length n formed from our collection by taking the
most common symbol at each position; the jth symbol of c therefore corresponds to the symbol 
having the maximum value in the j-th column of the profile matrix.
Of course, there may be more than one most common symbol, leading to multiple possible consensus strings.
'''

# Process Breakdown:
    # The code reads a file, removes certain characters from it and splits it into a list of strings
    # Then, it defines a function 'consensus' that receives a list of strings as input
    # The function calculates the frequency of each nucleotide in each position of the strings,
    # creates a consensus sequence based on the most frequent nucleotide in each position,
    # and returns the consensus sequence and the counts of each nucleotide in each position
    # Finally, the function writes the results to a file and closes it
    # The function is called with the list of strings as input, and its output is printed

'''
My summary:
    The first line opens a file called "Example12.txt" in read mode and assigns it to the variable 'f'
    The second line initializes an empty list called 'mat'
    The third line reads the entire contents of the file into a string called 'str1'
    The fourth line removes all instances of the string "Rosalind_" from 'str1'
    The fifth line removes all newline characters from 'str1'
    The sixth line removes all digits from 'str1'
    The seventh line splits 'str1' into a list of strings at each ">" character and assigns it to 'mat'
    The eighth line removes any empty strings from 'mat'
'''
f = open("C:/Users/varen/Desktop/My_New_Stuff/no_place_like/Computing/Bioinformatics/Rosalind/10-cons/Example12.txt", "r")
mat = []
str1 = f.read()
str1 = str1.replace("Rosalind_", "")
str1 = str1.replace("\n", "")
str1 = ''.join([i for i in str1 if not i.isdigit()])
mat = str1.split(">")
mat.remove("")

'''
The 'consensus' function takes a list of strings 'm' as input.
Within the function, four empty lists are initialized to store the count of each nucleotide at each position.
Two nested for-loops iterate over each position and each string in 'm', respectively.
For each string, the nucleotide at the current position is counted and added to the appropriate list.
'''
def consensus (m):
    a = []
    c = []
    g = []
    t = []
    for i in range(0, len(mat[0])):
        countA = 0
        countC = 0
        countG = 0
        countT = 0
        for j in mat:
            if (j[i] == "A"):
                countA = countA + 1
            elif (j[i] == "C"):
                countC = countC + 1
            elif (j[i] == "G"):
                countG = countG + 1
            elif (j[i] == "T"):
                countT = countT + 1
        '''
        The counts of each nucleotide at each position are converted to strings
        and concatenated with their respective nucleotide
        and a colon to create four strings 'strA', 'strC', 'strG', and 'strT'.
        '''
        a.append(countA)
        c.append(countC)
        g.append(countG)
        t.append(countT)
    profile = []
    for i in range(0, len(mat[0])):
        if (a[i] >= c[i] and a[i] >= g[i] and a[i] >= t[i]):
            profile.append("A")
        elif (c[i] >= a[i] and c[i] >= g[i] and c[i] >= t[i]):
            profile.append("C")
        elif (g[i] >= a[i] and g[i] >= c[i] and g[i] >= t[i]):
            profile.append("G")
        else:
            profile.append("T")
    strProfile = ''.join(profile)
    strA = 'A: ' + ' '.join(map(str,a))
    strC = 'C: ' + ' '.join(map(str,c))
    strG = 'G: ' + ' '.join(map(str,g))
    strT = 'T: ' + ' '.join(map(str,t))
    '''
    The five resulting strings ('strProfile', 'strA', 'strC', 'strG', and 'strT') are written to a file
    called "result12.txt".
    Finally, the function returns the five strings as a tuple.
    The function is called with 'mat' as input and its output is printed
    '''
    file = open("C:/Users/.../10-cons/result12.txt","w")
    file.write(strProfile)
    file.write("\n")
    file.write(strA)
    file.write("\n")
    file.write(strC)
    file.write("\n")
    file.write(strG)
    file.write("\n")
    file.write(strT)
    file.write("\n")
    file.close()
    return strProfile, strA, strC, strG, strT
print(consensus(mat))

File: 10-cons/cons.c
/*
#Problem 10: Consensus and Profile
#Given: A collection of at most 10 DNA strings of equal length (at most 1 kbp) in FASTA format.
#Return: A consensus string and profile matrix for the collection. (If several possible consensus strings exist, then you may return any one of them.)

'''
A matrix is a rectangular table of values divided into rows and columns. An m×n matrix has m rows and
n columns. Given a matrix A, we write Ai,j to indicate the value found at the
intersection of row i and column j.

Say that we have a collection of DNA strings, all having the same length n.
Their profile matrix is a 4×n matrix P in which P1,j represents the number of times
that 'A' occurs in the jth position of one of the strings, P2,j represents the number of times
that C occurs in the jth position, and so on (see below).

A consensus string c is a string of length n formed from our collection by taking the
most common symbol at each position; the jth symbol of c therefore corresponds to the symbol 
having the maximum value in the j-th column of the profile matrix.
Of course, there may be more than one most common symbol, leading to multiple possible consensus strings.
'''

# The code reads a file, removes certain characters from it and splits it into a list of strings.
# Then, it defines a function 'consensus' that receives a list of strings as input.
# The function calculates the frequency of each nucleotide in each position of the strings,
# creates a consensus sequence based on the most frequent nucleotide in each position,
# and returns the consensus sequence and the counts of each nucleotide in each position.
# Finally, the function writes the results to a file and closes it.
# The function is called with the list of strings as input, and its output is printed.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_DNA_STRINGS 10
#define MAX_DNA_LENGTH 1000

// Function to open the input file and extract the DNA strings
void parse_input_file(char* filename, char **dna_strings, int* num_strings, int* string_length) {
    FILE* f = fopen(filename, "r");
    if (f == NULL) {
        printf("Error opening file\n");
        exit(1);
    }

    // Read in the DNA strings from the FASTA file
    char line[MAX_DNA_LENGTH + 1];
    int i = 0;
    while (fgets(line, sizeof(line), f) != NULL) {
        if (line[0] == '>') {
            // Skip the header line
            continue;
        }
        // Remove newline characters from the end of the line
        line[strcspn(line, "\n")] = 0;
        // Copy the DNA string into the array
        strcpy(dna_strings[i], line);
        i++;
    }
    *num_strings = i;
    *string_length = strlen(dna_strings[0]);

    fclose(f);
}

// Function to create the profile matrix
void create_profile_matrix(char **dna_strings, int num_strings, int string_length, int **profile_matrix) {
    // Initialize the profile matrix to all zeros
    memset(*profile_matrix, 0, sizeof(int) * 4 * string_length);

    // Loop through each position in the DNA strings
    for (int i = 0; i < string_length; i++) {
        // Loop through each DNA string and count the occurrences of each nucleotide
        for (int j = 0; j < num_strings; j++) {
            char nucleotide = dna_strings[j][i];
            switch (nucleotide) {
                case 'A':
                    profile_matrix[i][0]++;
                    break;
                case 'C':
                    profile_matrix[i][1]++;
                    break;
                case 'G':
                    profile_matrix[i][2]++;
                    break;
                case 'T':
                    profile_matrix[i][3]++;
                    break;
                default:
                    printf("Invalid nucleotide: %c\n", nucleotide);
                    exit(1);
            }
        }
    }
}

// Function to generate the consensus string
void generate_consensus_string(int profile_matrix[][4], int string_length, char* consensus) {
    // Loop through each position in the profile matrix and find the most common nucleotide
    for (int i = 0; i < string_length; i++) {
        int max_count = 0;
        char max_nucleotide;
        for (int j = 0; j < 4; j++) {
            if (profile_matrix[i][j] > max_count) {
                max_count = profile_matrix[i][j];
                switch (j) {
                    case 0:
                        max_nucleotide = 'A';
                        break;
                    case 1:
                        max_nucleotide = 'C';
                        break;
                    case 2:
                        max_nucleotide = 'G';
                        break;
                    case 3:
                        max_nucleotide = 'T';
                        break;
                }
            }
        }
        consensus[i] = max_nucleotide;
    }
}

int main() {
    //char dna_strings[MAX_DNA_STRINGS][MAX_DNA_LENGTH + 1];
    char **dna_strings = (char **)malloc(MAX_DNA_STRINGS * sizeof(char *));
    for (int i = 0; i < MAX_DNA_STRINGS; i++) {
        dna_strings[i] = (char *)malloc((MAX_DNA_LENGTH + 1) * sizeof(char));
    }
    char *consensus = (char *)malloc((MAX_DNA_LENGTH + 1) * sizeof(char));
    //int profile_matrix[MAX_DNA_LENGTH][4];
    int **profile_matrix = (int **)calloc(MAX_DNA_LENGTH, sizeof(int *));
    for (int i = 0; i < MAX_DNA_LENGTH; i++) {
            profile_matrix[i] = (int *)calloc(4, sizeof(int));
        }
    int num_strings, string_length;

    // Parse the input file
    parse_input_file("input.fasta", dna_strings, &num_strings, &string_length);

    // Create the profile matrix
    create_profile_matrix(dna_strings, num_strings, string_length, profile_matrix);

    // Generate the consensus string
    generate_consensus_string(profile_matrix, string_length, consensus);

    // Print out the consensus string and the profile matrix
    printf("%s\n", consensus);
    printf("A:");
    for (int i = 0; i < string_length; i++) {
        printf(" %d", profile_matrix[i][0]);
    }
    printf("\nC:");
    for (int i = 0; i < string_length; i++) {
        printf(" %d", profile_matrix[i][1]);
    }
    printf("\nG:");
    for (int i = 0; i < string_length; i++) {
        printf(" %d", profile_matrix[i][2]);
    }
    printf("\nT:");
    for (int i = 0; i < string_length; i++) {
        printf(" %d", profile_matrix[i][3]);
    }
    printf("\n");

    for (int i = 0; i < MAX_DNA_STRINGS; i++) {
        free(dna_strings[i]);
    }
    free(dna_strings);
    for (int i = 0; i < string_length; i++) {
        free(profile_matrix[i]);
    }
    free(profile_matrix);

    return 0;
}


/*
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_DNA_STRINGS 10
#define MAX_DNA_LENGTH 1000

// Function to open the input file and extract the DNA strings
void parse_input_file(char* filename, char dna_strings[][MAX_DNA_LENGTH + 1], int* num_strings, int* string_length) {
    FILE* f = fopen(filename, "r");
    if (f == NULL) {
        printf("Error opening file\n");
        exit(1);
    }

    // Read in the DNA strings from the FASTA file
    char line[MAX_DNA_LENGTH + 1];
    int i = 0;
    while (fgets(line, sizeof(line), f) != NULL) {
        if (line[0] == '>') {
            // Skip the header line
            continue;
        }
        // Remove newline characters from the end of the line
        line[strcspn(line, "\n")] = 0;
        // Copy the DNA string into the array
        strcpy(dna_strings[i], line);
        i++;
    }

    *num_strings = i;
    *string_length = strlen(dna_strings[0]);

    fclose(f);
}

// Function to create the profile matrix
void create_profile_matrix(char dna_strings[][MAX_DNA_LENGTH + 1], int num_strings, int string_length, int profile_matrix[][4]) {
    // Initialize the profile matrix to all zeros
    memset(profile_matrix, 0, sizeof(int) * 4 * string_length);

    // Loop through each position in the DNA strings
    for (int i = 0; i < string_length; i++) {
        // Loop through each DNA string and count the occurrences of each nucleotide
        for (int j = 0; j < num_strings; j++) {
            char nucleotide = dna_strings[j][i];
            switch (nucleotide) {
                case 'A':
                    profile_matrix[i][0]++;
                    break;
                case 'C':
                    profile_matrix[i][1]++;
                    break;
                case 'G':
                    profile_matrix[i][2]++;
                    break;
                case 'T':
                    profile_matrix[i][3]++;
                    break;
                default:
                    printf("Invalid nucleotide: %c\n", nucleotide);
                    exit(1);
            }
        }
    }
}

// Function to generate the consensus string
void generate_consensus_string(int profile_matrix[][4], int string_length, char* consensus) {
    // Loop through each position in the profile matrix and find the most common nucleotide
    for (int i = 0; i < string_length; i++) {
        int max_count = 0;
        char max_nucleotide;
        for (int j = 0; j < 4; j++) {
            if (profile_matrix[i][j] > max_count) {
                max_count = profile_matrix[i][j];
                switch (j) {
                    case 0:
                        max_nucleotide = 'A';
                        break;
                    case 1:
                        max_nucleotide = 'C';
                        break;
                    case 2:
                        max_nucleotide = 'G';
                        break;
                    case 3:
                        max_nucleotide = 'T';
                        break;
                }
            }
        }
        consensus[i] = max_nucleotide;
    }
}

int main() {
    char dna_strings[MAX_DNA_STRINGS][MAX_DNA_LENGTH + 1];
    char consensus[MAX_DNA_LENGTH + 1];
    int profile_matrix[MAX_DNA_LENGTH][4];
    int num_strings, string_length;

    // Parse the input file
    parse_input_file("input.fasta", dna_strings, &num_strings, &string_length);

    // Create the profile matrix
    create_profile_matrix(dna_strings, num_strings, string_length, profile_matrix);

    // Generate the consensus string
    generate_consensus_string(profile_matrix, string_length, consensus);

    // Print out the consensus string and the profile matrix
    printf("%s\n", consensus);
    printf("A:");
    for (int i = 0; i < string_length; i++) {
        printf(" %d", profile_matrix[i][0]);
    }
    printf("\nC:");
    for (int i = 0; i < string_length; i++) {
        printf(" %d", profile_matrix[i][1]);
    }
    printf("\nG:");
    for (int i = 0; i < string_length; i++) {
        printf(" %d", profile_matrix[i][2]);
    }
    printf("\nT:");
    for (int i = 0; i < string_length; i++) {
        printf(" %d", profile_matrix[i][3]);
    }
    printf("\n");

    return 0;
}
*/